File: C11/백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100001

int N, K;
int dist[MAX];
int queue[MAX];
int front = 0, rear = 0;
int minTime = 1e9;
int count = 0;

void push(int x) {
    queue[rear++] = x;
}

int pop() {
    return queue[front++];
}

int empty() {
    return front == rear;
}

void bfs() {
    for (int i = 0; i < MAX; i++) dist[i] = -1;
    push(N);
    dist[N] = 0;

    while (!empty()) {
        int cur = pop();

        if (cur == K) {
            if (dist[cur] < minTime) {
                minTime = dist[cur];
                count = 1;
            } else if (dist[cur] == minTime) {
                count++;
            }
            continue;
        }

        int nexts[3] = {cur - 1, cur + 1, cur * 2};
        for (int i = 0; i < 3; i++) {
            int next = nexts[i];
            if (next < 0 || next >= MAX) continue;
            if (dist[next] == -1 || dist[next] == dist[cur] + 1) {
                dist[next] = dist[cur] + 1;
                push(next);
            }
        }
    }
}

int main() {
    scanf("%d %d", &N, &K);

    if (N >= K) {
        printf("%d\n1\n", N - K);
        return 0;
    }

    bfs();
    printf("%d\n%d\n", minTime, count);
    return 0;
}

File: Java/백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, K;
    static int[] dist = new int[100001];
    static int minTime = Integer.MAX_VALUE;
    static int count = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        if (N >= K) {
            System.out.println((N - K) + "\n1");
            return;
        }

        Arrays.fill(dist, -1);
        bfs();
        System.out.println(minTime);
        System.out.println(count);
    }

    static void bfs() {
        Queue<Integer> q = new LinkedList<>();
        q.add(N);
        dist[N] = 0;

        while (!q.isEmpty()) {
            int cur = q.poll();

            if (cur == K) {
                if (dist[cur] < minTime) {
                    minTime = dist[cur];
                    count = 1;
                } else if (dist[cur] == minTime) {
                    count++;
                }
                continue;
            }

            for (int next : new int[]{cur - 1, cur + 1, cur * 2}) {
                if (next < 0 || next > 100000) continue;

                if (dist[next] == -1 || dist[next] == dist[cur] + 1) {
                    dist[next] = dist[cur] + 1;
                    q.add(next);
                }
            }
        }
    }
}